### PR TITLE
Update release pipeline

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #4961

* Publish insiders daily at 2am PDT (9am UTC)
* Publish release

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -61,24 +63,37 @@ jobs:
...
         # We're not running CI on macOS for now because it's one less matrix entry to lower the number of runners used,
         # macOS runners are expensive, and we assume that Ubuntu is enough to cover the UNIX case.
         run: |
           echo "vsix_name=ms-toolsai-jupyter-insiders.vsix" >> $GITHUB_ENV
           echo "test_matrix_os=[\"ubuntu-latest\", \"windows-latest\"]" >> $GITHUB_ENV
+          echo "release_channel=insider" >> $GITHUB_ENV
+
+      - name: insiders channel
+        # Scheduled builds will publish Insider builds.
+        if: github.event_name == 'schedule' && github.ref == 'refs/heads/mains'
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:76]: We push insiders to marketplace only during the nightly run

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -61,24 +63,37 @@ jobs:
...
       - name: release
-        if: github.event_name == 'push' && github.ref != 'refs/heads/main'
+        if: github.event_name == 'push' && contains(github.ref, 'refs/heads/release')
         run: |
           echo "vsix_name=ms-toolsai-jupyter-release.vsix" >> $GITHUB_ENV
           echo "test_matrix_os=[\"ubuntu-latest\"]" >> $GITHUB_ENV
 
+      - name: release channel
+        # All pushes to release can push to marketplace place with the words `release` `publish` in commit.
+        if: github.event_name == 'push' && contains(github.ref, 'refs/heads/release') && contains(github.event.head_commit.message, 'release') && contains(github.event.head_commit.message, 'publish')
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:88]: We push insiders to marketplace only during push to release & commits has special comments

File: .github/workflows/build-test.yml [hunk 3]
```diff
@@ -1016,3 +1031,80 @@ jobs:
...
+        with:
+          node-version: ${{env.NODE_VERSION}}
+
+      - name: Download VSIX
+        uses: actions/download-artifact@v2
+        with:
+          name: ${{needs.pick_environment.outputs.vsix_name}}
+
+    #   - name: Publish
+    #     run: vsce publish --packagePath ${{needs.pick_environment.outputs.vsix_name}} --pat ${{secrets.VSCE_TOKEN}} --noVerify
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:1070]: Once reviewed, i'll enable this, only way to find out is to push a VSIX for insiders and test.

File: .github/workflows/build-test.yml [hunk 4]
```diff
@@ -1016,3 +1031,79 @@ jobs:
...
+        with:
+          node-version: ${{env.NODE_VERSION}}
+
+      - name: Download VSIX
+        uses: actions/download-artifact@v2
+        with:
+          name: ${{needs.pick_environment.outputs.vsix_name}}
+
+      - name: Publish
+        run: vsce publish --packagePath ${{needs.pick_environment.outputs.vsix_name}} --pat ${{secrets.VSCE_TOKEN}} --noVerify
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:1070]: Publish insider/release

File: .github/workflows/build-test.yml [hunk 5]
```diff
@@ -1016,3 +1031,79 @@ jobs:
...
+
+      - name: Extract Extension
+        if: needs.pick_environment.outputs.release_channel == 'stable'
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:1073]: All of the rest only applies when we generate stable releases

File: .github/workflows/build-test.yml [hunk 6]
```diff
@@ -19,8 +19,9 @@ on:
       - 'release'
       - 'release/*'
       - 'release-*'
+  schedule:
+    - cron: "0 9 * * *" 9am UTC (2am PDT)
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:23]: Nightly job

File: .github/workflows/build-test.yml [hunk 7]
```diff
@@ -1016,3 +1031,79 @@ jobs:
...
       #  run: bash <(curl -s https://codecov.io/bash) -f ./coverage/cobertura-coverage.xml -v
+
+  release:
+    name: Release
+    runs-on: ubuntu-latest
+    # This should happen in the release and insiders branch
+    if: github.repository == 'microsoft/vscode-jupyter' && (needs.pick_environment.outputs.release_channel == 'insider' || needs.pick_environment.outputs.release_channel == 'stable')
+    needs: [pick_environment, build-vsix, lint, ts_tests, smoke-tests]
+    environment:
+      name: ${{needs.pick_environment.outputs.release_channel}}
```
Comment (DonJayamanne) [inline .github/workflows/build-test.yml:1042]: Using Github Environments Releases
I have created two environments `insider` & `stable`.
For Insiders, they will get published directly.
For `stable` environment, this job will NOT run, until it is approved. there is a list of approvers that we can add (for now I had only added myself, to ensure this works properly, then we can add everyone else).

See here for more info:
https://docs.github.com/en/actions/reference/environments
https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments

**Review by rchiodo — APPROVED** [review]

File: .github/workflows/build-test.yml (hunk 7 above)
Comment (DonJayamanne) [reply .github/workflows/build-test.yml:1042]: Oh yes, I will add @IanMatthewHuff  as he's going to be doing the release this time

**Review by DavidKutu — APPROVED** [review]

**Pull request merged**